### PR TITLE
Fix some storyboard loops not playing for as long as expected

### LINKED ISSUE #23092: Storyboard not accounting for when loop events end
### Type

Game behaviour

### Bug description

In http://osu.ppy.sh/s/1388590, at around 1:54, the scene mimics a concert with lights on the sides. In stable this plays correctly, but in lazer, the highlights end too early (they just disappear)
This behavior is fixed when the sprite has an explicit end time, and not a loop event with just a start time

Lazer:
Screenshot 2023-04-02 143124

Stable:
image

Here's the .osb code for one of the disappearings
```Sprite,Background,Centre,"sb\hl.png",0,420
 L,99206,8
  F,12,0,176,1,0
  F,12,176,352,1,0
  F,12,352,529,1,0
  F,12,529,882,1,0
  F,12,882,1058,1,0
  F,12,1058,1235,1,0
  F,12,1235,1411,1,0
  F,12,1411,1764,1,0
  F,12,1764,1941,1,0
  F,12,1941,2117,1,0
  F,12,2117,2294,1,0
  F,12,2294,2824,1,0
 S,0,99206,,0.8931186
 C,0,99206,,220,57,38
```
As in the .osb code, there is no explicit end time, it needs to be calculated based on the given syntax in the loop.
If a command with the explicit end time of the sprite was added, this issue disappears.

### Screenshots or videos

_No response_

### Version

2023.403.0

### Logs

Logs aren't needed for this (I think?)

### PR DESCRIPTION
**Pull request by peppy**
Closes #23092.

A bit annoying that we need to add a separate definition for this, but I can't see a better way.

### CODE REVIEW

File: osu.Game/Storyboards/IStoryboardElementWithDuration.cs [hunk 1]
```diff
@@ -12,9 +12,17 @@ public interface IStoryboardElementWithDuration : IStoryboardElement
...
+        /// This is consumed to extend the length of a storyboard to ensure all visuals are played to completion.
         /// </summary>
         double EndTime { get; }
 
+        /// <summary>
+        /// The time this element displays until.
+        /// This is used for lifetime purposes, and includes long playing animations which don't necessarily extend
+        /// a storyboard's play time.
+        /// </summary>
+        double EndTimeForDisplay { get; }
```
Comment (bdach) [inline osu.Game/Storyboards/IStoryboardElementWithDuration.cs:24]: Not sure on the naming here... Thoughts on doing something like

- `EndTime` => `EarliestEndTime` (to reflect that it's the "earliest" time that the element is allowed to terminate being animated at)
- `EndTimeForDisplay` => `CompleteEndTime`?

Probably would need to adjust `StoryboardElementExtensions.GetEndTime()` to be called `GetEarliestEndTime()` instead or something.

File: osu.Game.Tests/Beatmaps/Formats/LegacyStoryboardDecoderTest.cs [hunk 2]
```diff
@@ -95,6 +95,27 @@ public void TestDecodeStoryboardEvents()
...
+                var storyboard = decoder.Decode(stream);
+
+                StoryboardLayer background = storyboard.Layers.Single(l => l.Depth == 3);
+                Assert.AreEqual(1, background.Elements.Count);
+
+                Assert.AreEqual(2000, background.Elements[0].StartTime);
+                Assert.AreEqual(2000, (background.Elements[0] as StoryboardAnimation)?.EarliestTransformTime);
+
+                Assert.AreEqual(3000, (background.Elements[0] as StoryboardAnimation)?.GetEndTime());
+                Assert.AreEqual(12000, (background.Elements[0] as StoryboardAnimation)?.EndTimeForDisplay);
```
Comment (bdach) [inline osu.Game.Tests/Beatmaps/Formats/LegacyStoryboardDecoderTest.cs:115]: I'm confused. This is fine and all, but the only thing that reads `EndTimeForDisplay` is... this test? How is this having any discernible effect anywhere in game? In fact, checking the beatmap linked in the issue thread, this PR doesn't seem to fix anything, visually speaking.

From the xmldoc I presume the property was also meant to be used here:

https://github.com/ppy/osu/blob/14af2063e5ec2abb16f751480f1702f54b6fe9ad/osu.Game/Storyboards/Drawables/DrawableStoryboardSprite.cs#L85

and maybe here?

https://github.com/ppy/osu/blob/14af2063e5ec2abb16f751480f1702f54b6fe9ad/osu.Game/Storyboards/Drawables/DrawableStoryboardAnimation.cs#L88

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game.Tests/Beatmaps/Formats/LegacyStoryboardDecoderTest.cs (hunk 2 above)
Comment (peppy) [reply osu.Game.Tests/Beatmaps/Formats/LegacyStoryboardDecoderTest.cs:115]: Seems to be a case of commit staging failure. I did test it to fix with the lines you propose, yes.

**Review by bdach — APPROVED** [review]
Let's go with this

**Pull request merged**